earn/src/components/borrow/modal/AddCollateralModal.tsx

```diff
@@ -0,0 +1,275 @@
+      setIsOpen(false);
+    } else if (!contractIsLoading && !contractDidSucceed) {
+      setIsPending(false);
+    }
+  }, [contractDidSucceed, contractData, contractIsLoading, setPendingTxn, setIsOpen]);
+
+  let confirmButtonState = ConfirmButtonState.READY;
+
+  if (collateralAmount.gt(userBalance)) {
+    confirmButtonState = ConfirmButtonState.INSUFFICIENT_ASSET;
+  } else if (isPending) {
+    confirmButtonState = ConfirmButtonState.PENDING;
+  }
+
+  const confirmButton = getConfirmButton(confirmButtonState, collateralToken);
+
+  return (
+    <FilledStylizedButton
+      size='M'
+      fillWidth={true}
+      disabled={!confirmButton.enabled}
+      onClick={() => {
+        if (confirmButtonState === ConfirmButtonState.READY) {
+          setIsPending(true);
+          contractWrite?.();
+        }
+      }}
+    >
+      {confirmButton.text}
+    </FilledStylizedButton>
+  );
+}
+
+export type AddCollateralModalProps = {
+  marginAccount: MarginAccount;
+  marketInfo: MarketInfo;
+  isOpen: boolean;
+  setIsOpen: (open: boolean) => void;
+  setPendingTxn: (pendingTxn: SendTransactionResult | null) => void;
+};
+
+export default function AddCollateralModal(props: AddCollateralModalProps) {
+  const { marginAccount, isOpen, setIsOpen, setPendingTxn } = props;
+  const { activeChain } = useContext(ChainContext);
+
+  const [collateralAmount, setCollateralAmount] = useState('');
+  const [collateralToken, setCollateralToken] = useState(marginAccount.token0);
+
+  const { address: userAddress } = useAccount();
+
+  const { refetch: refetchBalance, data: userBalance } = useBalance({
+    address: userAddress,
+    token: collateralToken.address,
+    chainId: activeChain.id,
+  });
+
+  useEffect(() => {
+    let interval: NodeJS.Timer | null = null;
+    if (isOpen) {
+      interval = setInterval(() => {
+        refetchBalance();
+      }, 13_000);
+    }
+    if (!isOpen && interval != null) {
+      clearInterval(interval);
+    }
+    return () => {
+      if (interval != null) {
+        clearInterval(interval);
+      }
+    };
+  }, [isOpen, refetchBalance]);
+
+  const resetModal = () => {
+    setCollateralAmount('');
+    setCollateralToken(marginAccount.token0);
+  };
+
+  const tokenOptions = useMemo(() => {
+    return [marginAccount.token0, marginAccount.token1];
+  }, [marginAccount.token0, marginAccount.token1]);
+
+  const existingCollateral =
+    collateralToken.address === marginAccount.token0.address
+      ? marginAccount.assets.token0Raw
+      : marginAccount.assets.token1Raw;
+
+  const numericCollateralAmount = Number(collateralAmount) || 0;
+
+  const collateralAmountBig = useMemo(() => {
+    return new Big(numericCollateralAmount).mul(10 ** collateralToken.decimals);
+  }, [collateralToken.decimals, numericCollateralAmount]);
+
+  const existingCollateralBig = useMemo(() => {
+    return new Big(existingCollateral).mul(10 ** collateralToken.decimals);
+  }, [collateralToken.decimals, existingCollateral]);
```

Comment:
Why `useMemo` for these? I don't see a reason for this simple calculation to be tied in with React's state management

Reply:
Good catch